### ISSUE TITLE
Base context v1 missing `sec:` mapping in RsaSignature2018 "@context"

### ISSUE THREAD

**Issue by kuzdogan** [w3c/vc-data-model#778]
I've been looking into the [base context](https://www.w3.org/2018/credentials/v1) and a minor PossibleErratum caught my attention. In the `"@context"` of `RsaSignature2018` property the shorthand `sec:` is used several times, but the definition is not given, as in other properties. 

RsaSignature2018:
```
"RsaSignature2018": {
      "@id": "https://w3id.org/security#RsaSignature2018",
      "@context": {
        "@version": 1.1,
        "@protected": true,

        "challenge": "sec:challenge",
        "created": {"@id": "http://purl.org/dc/terms/created", "@type": "xsd:dateTime"},
-->     "domain": "sec:domain",
        "expires": {"@id": "sec:expiration", "@type": "xsd:dateTime"},
-->     "jws": "sec:jws",
        "nonce": "sec:nonce",
        "proofPurpose": {
          "@id": "sec:proofPurpose",
          "@type": "@vocab",
          "@context": {
            "@version": 1.1,
            "@protected": true,

            "id": "@id",
            "type": "@type",

            "sec": "https://w3id.org/security#",

            "assertionMethod": {"@id": "sec:assertionMethod", "@type": "@id", "@container": "@set"},
            "authentication": {"@id": "sec:authenticationMethod", "@type": "@id", "@container": "@set"}
          }
        },
-->     "proofValue": "sec:proofValue",
        "verificationMethod": {"@id": "sec:verificationMethod", "@type": "@id"}
      }
    }
```
whereas in `EcdsaSecp256r1Signature2019` for example
<pre>
    "EcdsaSecp256r1Signature2019": {
      "@id": "https://w3id.org/security#EcdsaSecp256r1Signature2019",
      "@context": {
        "@version": 1.1,
        "@protected": true,

        "id": "@id",
        "type": "@type",

       <b> "sec": "https://w3id.org/security#",</b>
        "xsd": "http://www.w3.org/2001/XMLSchema#",

        "challenge": "sec:challenge",
        "created": {"@id": "http://purl.org/dc/terms/created", "@type": "xsd:dateTime"},
        "domain": "sec:domain",
        "expires": {"@id": "sec:expiration", "@type": "xsd:dateTime"},
        "jws": "sec:jws",
        "nonce": "sec:nonce",
        "proofPurpose": {
          "@id": "sec:proofPurpose",
          "@type": "@vocab",
          "@context": {
            "@version": 1.1,
            "@protected": true,

            "id": "@id",
            "type": "@type",

            "sec": "https://w3id.org/security#",

            "assertionMethod": {"@id": "sec:assertionMethod", "@type": "@id", "@container": "@set"},
            "authentication": {"@id": "sec:authenticationMethod", "@type": "@id", "@container": "@set"}
          }
        },
        "proofValue": "sec:proofValue",
        "verificationMethod": {"@id": "sec:verificationMethod", "@type": "@id"}
      }
    }
...
</pre>

**Comment by brentzundel** [comment]
I agree this is a correction that needs to be made to the security vocabulary. @OR13  @msporny @tplooker I believe you are the owners of the security vocab work item at the CCG.
We will close this issue as soon as it has been picked up by the security vocab folks.

**Comment by msporny** [comment]
> this is a correction that needs to be made to the security vocabulary.

Hrm, no, not in the security vocabulary... the bug is in the VC JSON-LD Context -- good catch @kuzdogan!

This means that RsaSignature2018 is completely broken in the 2018 Verifiable Credentials context... which tells me that no one is using that particular cryptosuite for VCs. :)

I'll note that the Ed25519Signature2018 definition doesn't have the same problem.

Changing this would be a substantive change... we will have to note that RsaSignature2018 MUST NOT be used with the 2018 VC suite. @kuzdogan -- how did you find this bug? Did a library throw an error on you?

**Comment by peacekeeper** [comment]
> which tells me that no one is using that particular cryptosuite for VCs. :)

I think one possible reason why this hasn't been detected sooner might be that during creation and verification of the `RsaSignature2018` proof, you never actually serialize the entire VC (including the proof) to N-Quads as it is shown above in a few examples.

You serialize the "proof options" separately from the "VC without the proof".

An implementation might serialize the "VC without the proof" using the `https://www.w3.org/2018/credentials/v1` context, and it might serialize the "proof options" using the `https://w3id.org/security/v2` context. Yes this would be a buggy implementation, but it's not inconceivable that the various CCG specs and discussions could be interpreted this way.

In such an implementation, the resulting proof would actually be correct, even though the VC as a whole would be invalid JSON-LD because of the missing terms.

**Comment by clehner** [comment]
> during creation and verification of the RsaSignature2018 proof, you never actually serialize the entire VC (including the proof) to N-Quads as it is shown above in a few examples.

This is correct. I've updated the examples I wrote to add the N-Quads for VC only and proof only.

For serializing the proof options, Spruce/[ssi](https://github.com/spruceid/ssi/)'s implementation merges the local context from the linked data document (VC without proof) into the proof object. So it still uses the compact IRIs in this case. If `https://w3id.org/security/v2` was used instead, that would result in the correct expanded IRIs: [On JSON-LD Playground](https://json-ld.org/playground/#startTab=tab-normalized&json-ld=%7B%22%40context%22%3A%22https%3A%2F%2Fw3id.org%2Fsecurity%2Fv2%22%2C%22type%22%3A%22RsaSignature2018%22%2C%22proofPurpose%22%3A%22assertionMethod%22%2C%22verificationMethod%22%3A%22did%3Aweb%3Ademo.spruceid.com%3A2021%3A05%3A10%3Aexample-rsa%23key%22%2C%22created%22%3A%222021-05-10T18%3A01%3A16.244Z%22%2C%22jws%22%3A%22eyJhbGciOiJSUzI1NiIsImtpZCI6InJzYTIwNDgtMjAyMC0wOC0yNSIsImNyaXQiOlsiYjY0Il0sImI2NCI6ZmFsc2V9..IM4BYrQ-1UIalVY35N9QaxNl4SXmvyniuWCMX1lGAd2O4wwqTd2TP3hZLjw4q_6rfqrCdhsk0fLdu3jfF1-RIaldZ_LdrqDyDSqSYlEGGn2XbUGFeFAwotRIW0Ay0ljah-FjRi62C487tgkbtP-S2Cr1yUNxZi6ZoGFqahEtUlDuFu0qIOTW4qtu_-w1JUEyr1FIN2Nfzt-F5dGst6WlieaizJHsS78pNnjZYRdI5jHig-gEUku7lA7yjivLsrLyQ_SOG0DgATkdR-vVWf8voxBiXL-ifMygCL7xyZduNpMfm5ADw2myhHKbDeT4EGKw1uMkV97ADVxknobsZUdTiQ%22%7D)

**Comment by kdenhartog** [comment]
I've optted to relabel this to a v2 issue that won't be addressed by this because of the reasons stated in this comment: https://github.com/w3c/vc-data-model/issues/778#issuecomment-836846943

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2022-08-03-vcwg#section6-9) on 2022-08-03

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 6.9. Base context v1 missing `sec:` mapping in RsaSignature2018 "@context" (issue vc-data-model#778)

_See github issue [vc-data-model#778](https://github.com/w3c/vc-data-model/issues/778)._




**Brent Zundel:** Not sure what this means.  

**Manu Sporny:** This deals with the base context... let's keep it in the data model since that's where the context is..  

> *Orie Steele:* +1 to keeping it, and then removing it..

**Brent Zundel:** Any opposition to keeping it in the core data model?.  

**Oliver Terbu:** Is this about the data integrity suite - RSA in particular?.  
… If so maybe would fit in Data Integrity.  

**Orie Steele:** I think there is a series of context-specific changes, like Mike Jones mentioned earlier..  
… Since that file is currently in the core data model repo, I think it's fine to leave issues in the core repo to say will keep or leave items from the core data model context..  
… This particular one I think we are definitely removing, and probably need to find in the data integrity repo at some point..  
… But I'm just guessing....  
… I suggest we leave issues that are requesting changes to the v1 context in the VC data repo until the change can be addressed by the WG..  

</details>

**Comment by msporny** [comment]
Listing specific Data Integrity cryptosuites have been removed from the v2 context and are going to be replaced with `DataIntegrityProof`. This issue is no longer relevant since the v2 context will not specifically list `RsaSignature2018`. Marking as pending close.

**Comment by brentzundel** [comment]
No comments since marked pending close, closing

**Comment by fabrii** [comment]
Hello!

I recently ran into a problem related to this. We are using verifiable-credentials-java (backend) and vc.js (frontend/mobile).
Both libraries handle this situation differently, and the canonicalization produced does not match. This means, as of now, that there is no interoperability between java and javascript for RsaSignature2018.

This type of error can be hard to diagnose, specially to newcommers as me. 

Also:

- RsaSignature2018 is used many times in https://www.w3.org/TR/vc-data-model/, and there is no warning of this.
- In our case, we have no possibility of changing the keys types away from RSA.
- Currently, there are not JsonWebSignature2020 javascript implementations that support RSA.

I believe that a consensus/workaround should be made explicit in the spec, so we can have interoperability across all implementations.

Thanks
